### PR TITLE
Parsing of gene annotations

### PR DESCRIPTION
**Pull request by joswinter**
Gene Annotations can now be loaded from the GUI and will be stored in the graph controller.

---

Will be merged at 11:59 14-06-2015

### CODE REVIEW

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParser.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+     * @param file
+     *            the file with gene annotations.
+     * @throws IOException
+     *             When there is an error reading the specified file.
+     * @return map from gene name to gene.
+     */
+    public static Map<String, GeneAnnotation> parseGeneAnnotations(
+            final File file) throws IOException {
+        Map<String, GeneAnnotation> genomeAnnotations = new HashMap<String, GeneAnnotation>();
+        Stream<String> annotationFile = Files.lines(file.toPath());
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParser.java]: `annotationLines` would be better name

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParser.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        annotationFile.forEach(line -> {
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParser.java]: To unleash the full lambda/stream power:

``` java
annotationFile.stream().map(GeneAnnotationParser::parseGeneAnnotation)
    .filter(genome -> genome != null)
    .foreach(genome -> genomeAnnotations.put(genome.getName(), genome));
```

**Comment by jorenham** [comment]
No Coveralls?

**Comment by rutgervandenberg** [comment]
Earlier check didn't include coveralls because of a http 500 while running the coverage test. I forced a rebuild and it works now. 

Code looks reasonable. There are a bunch of checkstyle violations, though they (may) not have been introduced in this pull request.

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 3]
```diff
@@ -241,6 +248,29 @@ private void initListeners() {
...
                         try {
+                            insertKnownMutations((File) args[0]);
+                        } catch (IOException exception) {
+                            shout(NotificationController.NOTIFY, "",
+                                    notFact.getNotification(exception));
+                        }
+                    }
+                });
+
+        listen(Message.OPENED,
```
Comment (alangerak) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java]: You don't need listen(Message.OPENED) 3 times, you can combine those to a single one.

**Comment by albert-smit** [comment]
looks good when above issues are fixed

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParser.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     *            the file with gene annotations.
+     * @throws IOException
+     *             When there is an error reading the specified file.
+     * @return map from gene name to gene.
+     */
+    public static Map<String, GeneAnnotation> parseGeneAnnotations(
+            final File file) throws IOException {
+        Map<String, GeneAnnotation> genomeAnnotations = new HashMap<String, GeneAnnotation>();
+        Stream<String> annotationLines = Files.lines(file.toPath());
+        annotationLines
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParser.java]: weird formatting

File: lifetiles-graph/src/test/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParserTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import org.junit.Test;
+
+public class GeneAnnotationParserTest {
+    static final String testGenomeFilename = "/data/test_annotations/simple_annotations";
+    private Map<String, GeneAnnotation> genomes;
+
+    @Before
+    public void setUp() throws Exception {
+        File genomeFile = new File(this.getClass()
+                .getResource(testGenomeFilename + ".gff").toURI());
```
Comment (jorenham) [inline lifetiles-graph/src/test/java/nl/tudelft/lifetiles/annotation/model/GeneAnnotationParserTest.java]: Why don't you add the extension directly?

**Comment by albert-smit** [comment]
updated version still contains multiple PMD and findbugs violations

```
[INFO] PMD Failure: nl.tudelft.lifetiles.annotation.model.GeneAnnotation:72 Rule:CommentRequired Priority:3 publicMethodCommentRequirement Required.

[INFO] PMD Failure: nl.tudelft.lifetiles.annotation.model.GeneAnnotationMapper:52 Rule:AvoidInstantiatingObjectsInLoops Priority:3 Avoid instantiating new objects inside loops.

[INFO] PMD Failure: nl.tudelft.lifetiles.annotation.model.GeneAnnotationParser:74 Rule:UseVarargs Priority:4 Consider using varargs for methods or constructors which take an array the last parameter..

[INFO] PMD Failure: nl.tudelft.lifetiles.annotation.model.KnownMutation:152 Rule:CommentRequired Priority:3 publicMethodCommentRequirement Required.

[INFO] PMD Failure: nl.tudelft.lifetiles.annotation.model.KnownMutationMapper:52 Rule:AvoidInstantiatingObjectsInLoops Priority:3 Avoid instantiating new objects inside loops.



[INFO] --- findbugs-maven-plugin:3.0.1:check (analyze-compile) @ lifetiles-graph ---

[INFO] BugInstance size is 2

[INFO] Error size is 0

[INFO] Total bugs: 2

[INFO] GeneAnnotation.mappingSegments not initialized in constructor and dereferenced in nl.tudelft.lifetiles.annotation.model.GeneAnnotation.getUnifiedPosition() [nl.tudelft.lifetiles.annotation.model.GeneAnnotation] At GeneAnnotation.java:[line 86]
```

**Comment by joswinter** [comment]
There are still stubs for the toCellString() method which are uncommented

**Comment by joswinter** [comment]
There was no implementation and no calls to the methods.

**Pull request merged**